initial draft of dump doc clarification

mikelle-rogers opened the pull request:
## Summary

Organize the documentation on collecting and analyzing dumps using .NET. Created a singular page that connects to all dump documentation. Fill in missing gaps.

Fixes #[2905](https://github.com/dotnet/diagnostics/issues/2905)

mikelle-rogers committed:
initial draft of dump doc clarification

mikelle-rogers committed:
adding .NET gcdump

mikelle-rogers committed:
fixed whitespace error and broken links

hoyosjs commented on docs/core/diagnostics/dumps.md:
```diff
@@ -20,57 +20,39 @@ Dumps can be collected in a variety of ways depending on which platform you are
 ## Analyze dumps
 
-You can analyze dumps using the [`dotnet-dump`](dotnet-dump.md) CLI tool or with [Visual Studio](/visualstudio/debugger/using-dump-files).
+You can analyze dumps using the [`dotnet-dump`](dotnet-dump.md) CLI tool on macOs, Linux and Windows. Linux and Windows dumps can be analyzed with [Visual Studio](/visualstudio/debugger/using-dump-files) and [Windbg](https://github.com/dotnet/docs/blob/main/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file) on Windows. LLDB is a tool that can analyze macOs and Linux dumps with bot ...
```
There's also the case where this is touching into when cross analysis is possible and when it's not. We have a "Analyze dumps collected on Linux" section, perhaps we should instead make this "Analyzing on Linux" and "Analyzing on Windows". To clarify and make that paragraph a little less dense we could split these few points into 

- dotnet-dump as the only fully way to analyze cross-os in the collection platform. Mention it only understands managed info.
- A section for analyzing on Windows. Mention VS and windbg being able to analyze native and managed state of Linux and Windows dumps. Also mention dotnet-dump on this platform is able to open Linux dumps
- A section on analyzing dumps on Linux. Say that lldb is the only debugger able to analyze native state on this platform, and that dotnet-sos extends its support into managed state with the example you gave.

mikelle-rogers replied:
The only thing I worry about there, is that people with a Linux dump won't be looking under the "analyzing on Windows" section. (Then, they will miss that they can analyze a Linux dump on windows)

mikelle-rogers replied:
I did something similar to what you suggested, what do you think?

hoyosjs reviewed (commented):
Missing a bit of the mem leak docs. I'll continue checking tomorrow - but some initial feedback. Starting to take shape

mikelle-rogers commented on docs/core/diagnostics/dumps.md:
```diff
@@ -20,57 +20,39 @@ Dumps can be collected in a variety of ways depending on which platform you are
 
 > [!NOTE]
-> Visual Studio version 16.8 and later allows you to [open Linux dumps](https://devblogs.microsoft.com/visualstudio/linux-managed-memory-dump-debugging/) generated on .NET Core 3.1.7 or later.
+> Visual Studio, on Windows, version 16.8 and later allows you to [open and analyze Linux dumps](https://devblogs.microsoft.com/visualstudio/linux-managed-memory-dump-debugging/) generated on .NET Core 3.1.7 or later.
```
What about VS for Mac?

hoyosjs replied:
The formal product name for that is Visual Studio for Mac.

mikelle-rogers commented on docs/core/diagnostics/dumps.md:
```diff
@@ -20,57 +20,39 @@ Dumps can be collected in a variety of ways depending on which platform you are
+### Collect dumps in a production environment or distributed system
 
-### Collect dumps at a specific point in time
+If you are running your app in production or you are running distributed systems, the best method to [collect a dump is through dotnet-monitor](dotnet-monitor.md).
```
What exactly do you mean by egress in this situation?

hoyosjs replied:
egress is how things leave a place. Usually in technical talk like this is the mechanisms supported to take the dump out of the machine or logs (e.g. in a blob store, to a database for logs, etc).

mikelle-rogers committed:
add platform specific files to create clarity

mikelle-rogers committed:
initial thoughts on memory analysis document organization

noahfalk commented on docs/core/diagnostics/memory-analysis.md:
```diff
@@ -0,0 +1,18 @@
+
+# Memory Analysis
+
+Various behaviors can be manifested by your app as a result of a memory leak. The most common are crashing, possibly with an out of memory exception, decrease of performance, consuming large amounts of cloud storage. The method to verify a memory leak as the problem is to investigate the memory usage of your app. Memory usage can be investigated at one specific point in time, or multiple points o ...
```
If the goal is just to address issue #2905 I don't think we need to add this new Memory Analysis page. If we are primarily going to talk about memory leaks then hopefully we could do that by editing the existing tutorial page for diagnosing memory leaks?

noahfalk commented on docs/core/diagnostics/debug-linux-dumps.md:
```diff
@@ -43,7 +43,10 @@ Once LLDB starts, it may be necessary to use the `setsymbolserver` command to po
 ## Analyze dumps on Windows
 
-Dumps collected from a Linux machine can also be analyzed on a Windows machine using [Visual Studio](/visualstudio/debugger/using-dump-files), [Windbg](/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file), or the [dotnet-dump](dotnet-dump.md) tool.
+Dumps collected from a Linux machine can also be analyzed on a Windows machine using [Visual Studio](/visualstudio/debugger/using-dump-files), [Windbg](/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file), or the [dotnet-dump](dotnet-dump.md) tool. Both Visual Studio and Windbg can analyze native and managed states, while dotnet-dump only understands managed information.
```
```suggestion
Dumps collected from a Linux machine can also be analyzed on a Windows machine using [Visual Studio](/visualstudio/debugger/using-dump-files), [Windbg](/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file), or the [dotnet-dump](dotnet-dump.md) tool. Both Visual Studio and Windbg can analyze native and managed code, while dotnet-dump only analyzes managed code.
```

noahfalk commented on docs/core/diagnostics/debug-windows-dumps.md:
```diff
@@ -0,0 +1,14 @@
+
+# Debug Windows dumps
+
+**This article applies to: ✔️** .NET Core 3.0 SDK and later versions
```
```suggestion
**This article applies to: ✔️** .NET Core 3.1 and later versions
```

noahfalk reviewed (changes requested):
I think we might be opening a can of worms if we encourage more people to use dotnet-gcdump and instead we should try to direct them away from it and make the tool obsolete. More comments inline.

mikelle-rogers commented on docs/core/diagnostics/dumps.md:
```diff
@@ -20,60 +20,39 @@ Dumps can be collected in a variety of ways depending on which platform you are
 
 ### Collect dumps on crash
```
Someone else said that this would be to much of a dense paragraph...

mikelle-rogers commented on docs/core/diagnostics/managed-memory-leak-gcdump.md:
```diff
@@ -0,0 +1,198 @@
+---
+title: Diagnose Managed Memory Leak with dotnet-gcdump
```
I did not mean to commit this file, sorry about that.

mikelle-rogers committed:
responding to code review

mikelle-rogers committed:
fixing whitespace

noahfalk reviewed (commented):
I'm not convinced the new memory-analysis.md page is an improvement over the existing diagnose memory leak tutorial, but the main worry I had about the gcdump tool page is resolved now : )

mikelle-rogers committed:
fix errors, clear memory-analysis

noahfalk commented on docs/core/diagnostics/dumps.md:
```diff
@@ -1,79 +1,47 @@
+* You may want to collect a dump when the app hasn't crashed yet. For example, if you want to examine the state of an application that seems to be in a deadlock, configuring the environment variables to collect dumps on crash will not be helpful because the app is still running.
 
-The following table shows the environment variables you can configure for collecting dumps on a crash.
+* [dotnet-dump](dotnet-dump.md) is a simple cross-platform command line tool to collect a dump. [dotnet-monitor] is an alternative if you need to collect a dump remotely or with automated triggering conditions. Several other debugger tools such as [Visual Studio](/visualstudio-docs/docs/debugger/using-dump-files) or [windbg](/windows-hardware/drivers/debugger/-dump--create-dump-file-) also have d ...
```
Nit: maybe we don't need to mention dotnet-monitor here if we are going to mention it again in the next bullet?

noahfalk commented on docs/core/diagnostics/dumps.md:
```diff
@@ -1,79 +1,47 @@
-> [!NOTE]
-> If native debugging is necessary, the [SOS debugger extension](sos-debugging-extension.md) can be used with [LLDB on Linux and macOS](debug-linux-dumps.md#analyze-dumps-on-linux). SOS is also supported with [Windbg/cdb](/windows-hardware/drivers/debugger/debugger-download-tools) on Windows, although Visual Studio is recommended.
+>[!NOTE]
+> [Visual Studio Memory Analysis](/visualstudio-docs/docs/profiling/analyze-memory-usage.md) can be used to diagnose a memory leak.
```
Using a NOTE gives me the idea that the VS option is perhaps 2nd class or less recommended. For folks on Windows I expect they will probably prefer VS. Can we change this to normal text and mention that it is for Windows only?

mikelle-rogers replied:
Can it find a memory leak in a Linux dump?

noahfalk replied:
Right now I don't think it can, but I asked the Sireesha on VS to be certain and I am waiting to hear back.

noahfalk commented on docs/core/diagnostics/debug-windows-dumps.md:
```diff
@@ -0,0 +1,14 @@
+
+## Analyze dumps on Windows
+
+Windows dumps can be analyzed on Windows with the following tools, [`dotnet-dump`](dotnet-dump.md) CLI tool, [Visual Studio](/visualstudio/debugger/using-dump-files), and [Windbg](https://github.com/dotnet/docs/blob/main/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file).
```
```suggestion
Windows dumps can be analyzed on Windows with [Visual Studio](/visualstudio/debugger/using-dump-files), [Windbg](https://github.com/dotnet/docs/blob/main/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file), or the [`dotnet-dump`](dotnet-dump.md) CLI tool.
```

I sorted them based on which on how easily I expect customers to find using them :)

noahfalk commented on docs/core/diagnostics/debug-memory-leak.md:
```diff
@@ -132,6 +132,9 @@ Once the dump is collected, you should have sufficient information to diagnose t
 In this tutorial, you're now done with the [Sample debug target](/samples/dotnet/samples/diagnostic-scenarios/) and you can close it. Navigate to the terminal that started the server, and press <kbd>Ctrl+C</kbd>.
 
+> [!NOTE]
+> After restarting the failed process and letting it run for around one minute, you could collect another dump in the same way explained above. You would then have two dumps over a period of time, and could compare them to assist in seeing where the memory is growing.
```
Collecting a dump after restart won't be very useful because restarting has reset all memory usage. For a more useful comparison over time they should let the original process continue running after collecting the first dump and collect the 2nd one prior to restarting. This NOTE should probably go at the end of the Generate Memory dump section.

noahfalk reviewed (approved):
A few small things inline but looks quite good now overall 👍 

Sorry for the slow review - I didn't realize you were waiting. Feel free to ping me directly anytime and let me know there is a review waiting. Usually I see the 'ready for review' email GitHub sends but not always.

mikelle-rogers committed:
replying to code review suggestions

mikelle-rogers committed:
order options for customer ease

noahfalk commented on docs/core/diagnostics/debug-memory-leak.md:
````diff
@@ -126,6 +126,9 @@ Writing minidump with heap to ./core_20190430_185145
 ```
 
+> [!NOTE]
+> For a comparison over time, let the original process continue running after collected the first dump and collect a second dump, as explained above, prior to restarting the app. You would then have two dumps over a period of time, and could compare them to assist in seeing where the memory is growing.
````
```suggestion
> For a comparison over time, let the original process continue running after collecting the first dump and collect a second dump the same way. You would then have two dumps over a period of time that you can compare to see where memory usage is growing.
```

noahfalk commented on docs/core/diagnostics/debug-windows-dumps.md:
```diff
@@ -0,0 +1,13 @@
+
+## Analyze dumps on Windows
+
+Windows dumps can be analyzed on Windows with [Visual Studio](/visualstudio/debugger/using-dump-files), and [Windbg](https://github.com/dotnet/docs/blob/main/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file), or the [`dotnet-dump`](dotnet-dump.md) CLI tool. Visual Studio and Windbg can access both managed and native code and are the recommended tools for a Windows environment. do ...
```
```suggestion
Windows dumps can be analyzed on Windows with [Visual Studio](/visualstudio/debugger/using-dump-files), [Windbg](https://github.com/dotnet/docs/blob/main/windows-hardware/drivers/debugger/analyzing-a-user-mode-dump-file), or the [`dotnet-dump`](dotnet-dump.md) CLI tool. Visual Studio and Windbg can access both managed and native code and are the recommended tools for a Windows environment. dotnet-dump can only access managed code.
```

noahfalk reviewed (approved):
A couple suggestions to simplify wording inline

mikelle-rogers committed:
clearing up time comparison

gewarren reviewed (changes requested):
And add the new files to the TOC at https://github.com/dotnet/docs/blob/main/docs/toc/tools-diagnostics/toc.yml.

mikelle-rogers commented on docs/core/diagnostics/collect-dumps-crash.md:
```diff
@@ -0,0 +1,43 @@
+|Environment variable|Description|Default value|
+|-------|---------|---|
+|`COMPlus_DbgEnableMiniDump` or `DOTNET_DbgEnableMiniDump`|If set to 1, enable core dump generation.|0|
+|`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|
```
It turns into a table: 
<img width="818" alt="image" src="https://user-images.githubusercontent.com/45022607/215917461-a43e910e-0e9b-4419-acad-e4d407202455.png">

gewarren replied:
Are you referring to the next row then?

mikelle-rogers replied:
Row 13 is the start of the table I am referring to.

mikelle-rogers replied:
It is referring to the minidump table below the table that the environment variables are in.

mikelle-rogers commented:
> And add the new files to the TOC at https://github.com/dotnet/docs/blob/main/docs/toc/tools-diagnostics/toc.yml.

I can only find docs/toc.yml which does not look like the correct file to add things to.

gewarren commented:
> > And add the new files to the TOC at https://github.com/dotnet/docs/blob/main/docs/toc/tools-diagnostics/toc.yml.
> 
> I can only find docs/toc.yml which does not look like the correct file to add things to. @gewarren

Sorry, it moved to https://github.com/dotnet/docs/blob/main/docs/navigate/tools-diagnostics/toc.yml.

mikelle-rogers committed:
Merge branch 'dotnet:main' into dev/mirogers/dumpdocs_

mikelle-rogers committed:
Adding to toc.yml- docs/navigate/tools-diagnostics

Adding new files to the table of contents. The new files are Debug Windows dumps and Collect dumps on crash.

mikelle-rogers commented:
> Sorry, it moved to https://github.com/dotnet/docs/blob/main/docs/navigate/tools-diagnostics/toc.yml.

I added the new files!

gewarren commented on docs/core/diagnostics/collect-dumps-crash.md:
```diff
@@ -0,0 +1,43 @@
+|Environment variable|Description|Default value|
+|-------|---------|---|
+|`COMPlus_DbgEnableMiniDump` or `DOTNET_DbgEnableMiniDump`|If set to 1, enable core dump generation.|0|
+|`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|
```
```suggestion
|`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see [Types of mini dumps](#types-of-mini-dumps).|2 (`MiniDumpWithPrivateReadWriteMemory`)|
```

gewarren replied:
I'm also confused why 2 corresponds to `MiniDumpWithPrivateReadWriteMemory` here, but in the referenced table, it corresponds to `Heap`. Are there two different ways to write it?

mikelle-rogers replied:
@noahfalk would you please lend some insight to this?

noahfalk replied:
Those two names are refering to the same concept and I'd recommend we replace MiniDumpWithPrivateReadWriteMemory -> Heap for naming consistency.

I believe what happened is in the past the doc used those very long names everywhere which correspond to some constants defined on the win32 MiniDumpWriteDump API. However someone probably recognized there was little value in using complicated API naming details when explaining these concepts. When they refactored to simpler names this reference probably got overlooked.

mikelle-rogers committed:
File Path templates

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

mikelle-rogers committed:
Types of minidump header

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

mikelle-rogers committed:
Link to table

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

mikelle-rogers committed:
Simplify name to heap

gewarren commented on docs/core/diagnostics/collect-dumps-crash.md:
```diff
@@ -0,0 +1,47 @@
+|`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see [Types of mini dumps](#types-of-mini-dumps).|2 (`Heap`)|
+|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to. Ensure that the user under which the dotnet process is running has write permissions to the specified directory.|`/tmp/coredump.<pid>`|
+|`COMPlus_CreateDumpDiagnostics` or `DOTNET_CreateDumpDiagnostics`|If set to 1, enables diagnostic logging of dump process.|0|
+|`COMPlus_EnableCrashReport` or `DOTNET_EnableCrashReport`|(Requires .NET 6 or later, not supported on Windows) If set to 1, the runtime generates a JSON-formatted crash report that includes information about the threads and stack frames of the crashing application. The crash report name is the dump path or name with *.crashreport.json* appended.
```
```suggestion
|`COMPlus_EnableCrashReport` or `DOTNET_EnableCrashReport`|(Requires .NET 6 or later; not supported on Windows.)<br/>If set to 1, the runtime generates a JSON-formatted crash report that includes information about the threads and stack frames of the crashing application. The crash report name is the dump path or name with *.crashreport.json* appended.
```

gewarren commented on docs/core/diagnostics/collect-dumps-crash.md:
```diff
@@ -0,0 +1,47 @@
+|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to. Ensure that the user under which the dotnet process is running has write permissions to the specified directory.|`/tmp/coredump.<pid>`|
+|`COMPlus_CreateDumpDiagnostics` or `DOTNET_CreateDumpDiagnostics`|If set to 1, enables diagnostic logging of dump process.|0|
+|`COMPlus_EnableCrashReport` or `DOTNET_EnableCrashReport`|(Requires .NET 6 or later, not supported on Windows) If set to 1, the runtime generates a JSON-formatted crash report that includes information about the threads and stack frames of the crashing application. The crash report name is the dump path or name with *.crashreport.json* appended.
+|`COMPlus_CreateDumpVerboseDiagnostics` or `DOTNET_CreateDumpVerboseDiagnostics`|(Requires .NET 7 or later) If set to 1, enables verbose diagnostic logging of the dump process.|0|
```
```suggestion
|`COMPlus_CreateDumpVerboseDiagnostics` or `DOTNET_CreateDumpVerboseDiagnostics`|(Requires .NET 7 or later.)<br/>If set to 1, enables verbose diagnostic logging of the dump process.|0|
```

gewarren commented on docs/core/diagnostics/collect-dumps-crash.md:
```diff
@@ -0,0 +1,47 @@
+|`COMPlus_CreateDumpDiagnostics` or `DOTNET_CreateDumpDiagnostics`|If set to 1, enables diagnostic logging of dump process.|0|
+|`COMPlus_EnableCrashReport` or `DOTNET_EnableCrashReport`|(Requires .NET 6 or later, not supported on Windows) If set to 1, the runtime generates a JSON-formatted crash report that includes information about the threads and stack frames of the crashing application. The crash report name is the dump path or name with *.crashreport.json* appended.
+|`COMPlus_CreateDumpVerboseDiagnostics` or `DOTNET_CreateDumpVerboseDiagnostics`|(Requires .NET 7 or later) If set to 1, enables verbose diagnostic logging of the dump process.|0|
+|`COMPlus_CreateDumpLogToFile` or `DOTNET_CreateDumpLogToFile`|(Requires .NET 7 or later) The path of the file to which the diagnostic messages should be written. If unset, the diagnostic messages are written to the console of the crashing application.|
```
```suggestion
|`COMPlus_CreateDumpLogToFile` or `DOTNET_CreateDumpLogToFile`|(Requires .NET 7 or later.)<br/>The path of the file to which the diagnostic messages should be written. If unset, the diagnostic messages are written to the console of the crashing application.|
```

gewarren commented on docs/core/diagnostics/collect-dumps-crash.md:
```diff
@@ -0,0 +1,47 @@
+
+## Types of mini dumps
+
+The following table shows all the values you can use for `DOTNET_DbgMiniDumpType`. For example, setting `DOTNET_DbgMiniDumpType` to 1 means `MiniDumpNormal` type dump will be collected on a crash.
```
```suggestion
The following table shows all the values you can use for `DOTNET_DbgMiniDumpType`. For example, setting `DOTNET_DbgMiniDumpType` to 1 means `Mini` type dump will be collected on a crash.
```

gewarren reviewed (approved):
Just a few final nits.

gewarren committed:
Apply suggestions from code review
